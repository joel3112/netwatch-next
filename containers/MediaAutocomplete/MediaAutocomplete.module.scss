.wrapper {
  display: block;
  height: 100%;
  position: relative;
}

.autocompleteInput {
  max-width: 100vw;
  width: 100%;
  height: 60px;
  padding-left: 10px;
  border-bottom: 2px solid var(--color-text);

  .input {
    font-size: rem(18px);
    font-weight: 500;
    padding-left: 18px;
    border-radius: 0;
    border: none;
  }
}

.autocompleteResult {
  box-shadow: var(--box-shadow-primary);
  margin-top: 10px;
}

.autocompleteItem {
  background-color: var(--background-color);
  border-bottom: 1px solid var(--color-disabled);
  display: flex;
  gap: 20px;
  color: var(--color-text);
  width: 100%;
  padding: 10px 15px;
  position: relative;

  &.selected,
  &:hover {
    background-color: rgb(var(--rgb-background-color-primary-hover) / 70%);
    color: var(--color-button-text);
  }

  .image {
    width: 50px;
    min-width: 50px;
  }

  .name {
    font-size: 0.91rem;
    font-weight: 500;
    margin-top: 10px;
  }

  .date {
    font-size: 0.85rem;
  }
}

.enter {
  position: absolute;
  top: 50%;
  right: 15px;
  transform: translateY(-50%);
  font-size: 1.5rem;
  font-weight: 300;
  display: none;

  .autocompleteItem.selected &,
  .autocompleteItem:hover & {
    display: flex;
  }
}

.message {
  @include absolute-center('horizontal');

  top: 110px;
}
