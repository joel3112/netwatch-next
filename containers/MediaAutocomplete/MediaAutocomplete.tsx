/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable jsx-a11y/interactive-supports-focus */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import { useState, useRef, useEffect, useMemo, ChangeEvent } from 'react';
import axios from 'axios';
import { AutocompleteState, createAutocomplete, GetSources } from '@algolia/autocomplete-core';
import Link from 'next/link';
import { FiSearch } from 'react-icons/fi';
import { BsArrowReturnLeft } from 'react-icons/bs';
import { FunctionGeneric, MediaData } from '@/types';
import { useI18n } from '@/hooks/useI18n';
import { Input } from '@/components/forms';
import { Image } from '@/components/media';
import { Space } from '@/components/layout';
import { classes } from '@/utils/helpers';
import styles from '@/containers/MediaAutocomplete/MediaAutocomplete.module.scss';

function debouncePromise(fn: FunctionGeneric, time: number) {
  let timerId: unknown = undefined;

  return function debounced(...args: unknown[]) {
    if (timerId) {
      clearTimeout(timerId as number);
    }

    return new Promise((resolve) => {
      timerId = setTimeout(() => resolve(fn(...args)), time);
    });
  };
}

const debounced = debouncePromise((items: unknown) => Promise.resolve(items), 400);

/* -------------------------------------------------------------------------- */
/** MediaAutocompleteItem (child component) **/
/* -------------------------------------------------------------------------- */

type MediaAutocompleteItemProps = {
  selected: boolean;
  href: string;
  name: string;
  image: string;
  date: string;
  onClick: () => void;
};

const MediaAutocompleteItem = ({
  selected,
  href,
  name,
  image,
  date,
  onClick
}: MediaAutocompleteItemProps) => {
  const handleClick = () => onClick && onClick();

  return (
    <Link href={href}>
      <a
        role="button"
        className={classes(styles.autocompleteItem, selected && styles.selected)}
        onClick={handleClick}>
        <div className={styles.image}>
          <Image src={image} alt="item" ratio={1.5} />
        </div>

        <Space direction="column" gap={5}>
          <span className={styles.name}>{name.truncate(70)}</span>
          {date && <span className={styles.date}>{date}</span>}
          <div className={styles.enter}>
            <BsArrowReturnLeft />
          </div>
        </Space>
      </a>
    </Link>
  );
};

/* -------------------------------------------------------------------------- */
/** MediaAutocomplete (main component) **/
/* -------------------------------------------------------------------------- */

export type MediaAutocompleteProps = typeof defaultProps & {
  onClose?: () => void;
};

const defaultProps = {};

type MediaAutocompleteState = Partial<AutocompleteState<MediaData>> & {
  isLoading: boolean;
};

const MediaAutocomplete = ({ onClose }: MediaAutocompleteProps) => {
  const truncateLength = 5;
  const formRef = useRef(null);
  const inputRef = useRef(null);
  const panelRef = useRef(null);

  const { t } = useI18n();
  const [autocompleteState, setAutocompleteState] = useState<MediaAutocompleteState>({
    isOpen: false,
    isLoading: false
  });

  useEffect(() => {
    setTimeout(() => {
      if (inputRef.current) {
        (inputRef.current as HTMLInputElement).focus();
      }
    }, 100);
  }, []);

  useEffect(() => {
    if (Number(autocompleteState.activeItemId) >= truncateLength) {
      setAutocompleteState((prev) => ({
        ...prev,
        activeItemId: Number(prev.activeItemId) % truncateLength
      }));
    }
  }, [autocompleteState.activeItemId]);

  const getSources = () => {
    return debounced([
      {
        sourceId: 'results',
        getItemUrl: ({ item }: { item: MediaData }) => `/${item.type}/${item.id}`,
        getItems: ({ query }: { query: string }) => {
          setAutocompleteState((prev) => ({ ...prev, isLoading: true }));

          return axios.get(`/api/search?query=${query}`).then(({ data }) => {
            setAutocompleteState((prev) => ({ ...prev, isLoading: false }));
            return data.results;
          });
        }
      }
    ]);
  };

  const autocomplete = useMemo(
    () =>
      createAutocomplete<MediaData, ChangeEvent<HTMLInputElement>>({
        id: 'autocomplete',
        placeholder: t('header.action.search.placeholder'),
        onStateChange: ({ state }) => {
          setAutocompleteState(state as MediaAutocompleteState);
        },
        autoFocus: true,
        getSources: getSources as unknown as GetSources<MediaData>
      }),
    [t]
  );

  const formProps = autocomplete.getFormProps({
    inputElement: inputRef.current
  });
  const inputProps = autocomplete.getInputProps({
    icon: FiSearch,
    inputElement: inputRef.current
  });

  const InputSearch = () => {
    return (
      <div className={classes(styles.autocompleteInput)}>
        <Input name="search" ref={inputRef} className={styles.input} {...inputProps} />
      </div>
    );
  };

  const handleItemSelect = () => {
    setAutocompleteState((prev) => ({ ...prev, isOpen: false }));
    onClose && onClose();
  };

  return (
    <div className={classes(styles.wrapper)}>
      <form className={styles.autocompleteForm} ref={formRef} {...(formProps as unknown)}>
        {InputSearch()}

        {autocompleteState.status === 'loading' && (
          <div className={styles.message}>{t('header.action.search.loading')}</div>
        )}

        {autocompleteState.status === 'idle' &&
        (!autocompleteState.collections?.length ||
          !autocompleteState.collections[0].items?.length) ? (
          <div className={styles.message}>{t('header.action.search.no.results')}</div>
        ) : (
          ''
        )}

        {autocompleteState.status === 'idle' && autocompleteState.collections?.length ? (
          <div ref={panelRef} {...(autocomplete.getPanelProps() as unknown)}>
            {autocompleteState.collections &&
              autocompleteState.collections.map((collection, index) => {
                const { items } = collection;

                return (
                  <section key={`section-${index}`}>
                    {items.length > 0 && (
                      <ul className={styles.autocompleteResult} {...autocomplete.getListProps()}>
                        {items.truncate(truncateLength).map((item: MediaData, index: number) => {
                          const id = `autocomplete-item-${index}`;

                          return (
                            <li
                              key={item.id}
                              id={id}
                              onClick={onClose}
                              onMouseEnter={() =>
                                setAutocompleteState((prev) => ({
                                  ...prev,
                                  activeItemId: index
                                }))
                              }>
                              <MediaAutocompleteItem
                                href={`/${item.type}/${item.id}`}
                                selected={autocompleteState.activeItemId === index}
                                onClick={handleItemSelect}
                                name={item.name}
                                image={item.image}
                                date={item.date || ''}
                              />
                            </li>
                          );
                        })}
                      </ul>
                    )}
                  </section>
                );
              })}
          </div>
        ) : (
          ''
        )}
      </form>
    </div>
  );
};

MediaAutocomplete.defaultProps = defaultProps;

export default MediaAutocomplete;
