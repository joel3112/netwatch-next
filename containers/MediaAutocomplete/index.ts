export { default as MediaAutocomplete } from '@/containers/MediaAutocomplete/MediaAutocomplete';
