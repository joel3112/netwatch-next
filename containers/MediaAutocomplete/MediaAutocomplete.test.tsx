import { render } from '@testing-library/react';
import MediaAutocomplete from '@/containers/MediaAutocomplete/MediaAutocomplete';

describe('Tests MediaAutocomplete component', () => {
  test('renders component correctly', () => {
    const { container } = render(<MediaAutocomplete />);

    expect(container).toMatchSnapshot();
  });
});
