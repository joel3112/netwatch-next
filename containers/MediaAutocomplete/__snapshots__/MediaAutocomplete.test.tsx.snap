// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests MediaAutocomplete component renders component correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <form
      action=""
      class="autocompleteForm"
      novalidate=""
      role="search"
    >
      <div
        class="autocompleteInput"
      >
        <div
          class="wrapper wrapper input"
        >
          <button
            aria-label="icon"
            class="wrapper medium outline icon"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="11"
                cy="11"
                r="8"
              />
              <line
                x1="21"
                x2="16.65"
                y1="21"
                y2="16.65"
              />
            </svg>
          </button>
          <input
            aria-label="search"
            autocomplete="off"
            class="input"
            name="search"
            placeholder="header.action.search.placeholder"
            type="search"
            value=""
          />
          <style>
            
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          
          
          align-items: center;
        }
      
          </style>
        </div>
      </div>
    </form>
  </div>
</div>
`;
